toolkit/tools/pkg/imagecustomizerlib/shrinkfilesystems.go

```diff
@@ -0,0 +1,151 @@
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+func shrinkFilesystems(imageLoopDevice string, outputImageFile string) error {
+	logger.Log.Infof("Shrinking filesystems")
+
+	// Get partition info
+	diskPartitions, err := diskutils.GetDiskPartitions(imageLoopDevice)
+	if err != nil {
+		return err
+	}
+
+	// Get the start sectors of all partitions
+	matchStarts, err := getStartSectors(imageLoopDevice, len(diskPartitions)-1)
+	if err != nil {
+		return err
+	}
+
+	for partitionNum := 0; partitionNum < len(diskPartitions); partitionNum++ {
+		if diskPartitions[partitionNum].Type == "part" {
+			fstype := diskPartitions[partitionNum].FileSystemType
+			// Currently only support ext2, ext3, ext4 filesystem types
+			if fstype != "ext2" && fstype != "ext3" && fstype != "ext4" {
```

Comment:
nit: if this check is being repeated, we can have a helper function to check for it...